File: web/src/denoise.ts
/** GPU à-trous denoising. Mirrors `crates/gpu/src/denoise.rs`.
 *
 * Runs over the finished accumulation buffer into a separate output shaped like
 * `Accum`, so the display pass reads it with no change at all — switching the
 * denoiser on swaps which buffer is bound and nothing else.
 *
 * The accumulation is never written. Denoising is a display-time product: one
 * more sample still converges to the unbiased answer.
 */
import { ACCUM_BYTES_PER_PIXEL } from './generated/layout';

export interface DenoiseSettings {
  /** À-trous passes. Each doubles the tap spacing. Zero disables the filter. */
  iterations: number;
  sigmaNormal: number;
  sigmaDepth: number;
  sigmaLuminance: number;
  sigmaAlbedo: number;
}

/** Matches `DenoiseParams::default()` in crates/core/src/denoise.rs. */
export const DEFAULT_DENOISE: DenoiseSettings = {
  iterations: 4,
  sigmaNormal: 64,
  sigmaDepth: 0.08,
  sigmaLuminance: 4,
  sigmaAlbedo: 0.1,
};

const PARAMS_BYTES = 32;

export class Denoiser {
  private layout: GPUBindGroupLayout;
  private demodulate: GPUComputePipeline;
  private measure: GPUComputePipeline;
  private atrous: GPUComputePipeline;
  private modulate: GPUComputePipeline;

  private a?: GPUBuffer;
  private b?: GPUBuffer;
  private deviation?: GPUBuffer;
  private params: GPUBuffer[] = [];
  private pixels = 0;

  /** Accum-shaped; what the display pass binds when denoising is on. */
  output?: GPUBuffer;

  private constructor(
    private device: GPUDevice,
    layout: GPUBindGroupLayout,
    pipelines: GPUComputePipeline[],
  ) {
    this.layout = layout;
    [this.demodulate, this.measure, this.atrous, this.modulate] = pipelines;
  }

  static async create(device: GPUDevice, module: GPUShaderModule): Promise<Denoiser> {
    const sto = (binding: number, readOnly: boolean): GPUBindGroupLayoutEntry => ({
      binding,
      visibility: GPUShaderStage.COMPUTE,
      buffer: { type: readOnly ? 'read-only-storage' : 'storage' },
    });
    const layout = device.createBindGroupLayout({
      label: 'denoise',
      entries: [
        { binding: 0, visibility: GPUShaderStage.COMPUTE, buffer: { type: 'uniform' } },
        sto(1, true),
        sto(2, false),
        sto(3, false),
        sto(4, false),
        sto(5, false),
      ],
    });
    const pl = device.createPipelineLayout({ bindGroupLayouts: [layout] });
    const stage = (entryPoint: string) =>
      device.createComputePipelineAsync({ label: entryPoint, layout: pl, compute: { module, entryPoint } });
    const pipelines = await Promise.all([
      stage('demodulate'),
      stage('measure'),
      stage('atrous_pass'),
      stage('modulate'),
    ]);
    return new Denoiser(device, layout, pipelines);
  }

  /** Reallocate for a new render size. */
  resize(pixels: number): void {
    if (pixels === this.pixels) return;
    for (const b of [this.a, this.b, this.deviation, this.output]) b?.destroy();
    const mk = (label: string, bytes: number) =>
      this.device.createBuffer({ label, size: bytes, usage: GPUBufferUsage.STORAGE });
    this.a = mk('denoise a', pixels * 16);
    this.b = mk('denoise b', pixels * 16);
    this.deviation = mk('denoise deviation', pixels * 4);
    this.output = mk('denoise output', pixels * ACCUM_BYTES_PER_PIXEL);
    this.pixels = pixels;
  }

  private paramsFor(index: number, width: number, height: number, stride: number, s: DenoiseSettings): GPUBuffer {
    while (this.params.length <= index) {
      this.params.push(
        this.device.createBuffer({
          label: 'denoise params',
          size: PARAMS_BYTES,
          usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
        }),
      );
    }
    const buf = new ArrayBuffer(PARAMS_BYTES);
    new Uint32Array(buf, 0, 3).set([width, height, stride]);
    new Float32Array(buf, 12, 4).set([s.sigmaNormal, s.sigmaDepth, s.sigmaLuminance, s.sigmaAlbedo]);
    this.device.queue.writeBuffer(this.params[index], 0, buf);
    return this.params[index];
  }

  /** Record the filter chain. Returns the buffer the display should read. */
  run(
    encoder: GPUCommandEncoder,
    accum: GPUBuffer,
    width: number,
    height: number,
    settings: DenoiseSettings,
  ): GPUBuffer {
    this.resize(width * height);
    const gx = Math.ceil(width / 8);
    const gy = Math.ceil(height / 8);

    const group = (u: GPUBuffer, a: GPUBuffer, b: GPUBuffer) =>
      this.device.createBindGroup({
        layout: this.layout,
        entries: [
          { binding: 0, resource: { buffer: u } },
          { binding: 1, resource: { buffer: accum } },
          { binding: 2, resource: { buffer: a } },
          { binding: 3, resource: { buffer: b } },
          { binding: 4, resource: { buffer: this.deviation! } },
          { binding: 5, resource: { buffer: this.output! } },
        ],
      });
    const pass = encoder.beginComputePass({ label: 'denoise' });
    const dispatch = (p: GPUComputePipeline, g: GPUBindGroup) => {
      pass.setPipeline(p);
      pass.setBindGroup(0, g);
      pass.dispatchWorkgroups(gx, gy, 1);
    };

    const g0 = group(this.paramsFor(0, width, height, 1, settings), this.a!, this.b!);
    dispatch(this.demodulate, g0);
    dispatch(this.measure, g0);

    // Ping-pong by swapping which buffer is bound where, so the shader always
    // reads `buf_a` and writes `buf_b` and needs no notion of which pass it is
    // on.
    let inA = true;
    for (let level = 0; level < settings.iterations; level++) {
      const u = this.paramsFor(level + 1, width, height, 1 << level, settings);
      const g = inA ? group(u, this.a!, this.b!) : group(u, this.b!, this.a!);
      dispatch(this.atrous, g);
      inA = !inA;
    }

    // `modulate` reads `buf_a`, so bind whichever buffer the last pass wrote.
    const uf = this.paramsFor(settings.iterations + 1, width, height, 1, settings);
    dispatch(this.modulate, inA ? group(uf, this.a!, this.b!) : group(uf, this.b!, this.a!));
    pass.end();
    return this.output!;
  }
}

File: crates/gpu/src/denoise.rs
//! GPU à-trous denoising. Mirrors `crates/core/src/denoise.rs`.
//!
//! Runs as its own chain of compute passes over the accumulation buffer,
//! writing a separate output the display pass reads. The accumulation is never
//! touched — denoising is a display-time product, and one more sample still
//! converges to the unbiased answer.
//!
//! Its own bind group, so the eight-storage-buffer limit that shapes everything
//! else in this renderer is not a constraint here: the tracing kernels are not
//! involved.

use crate::{Gpu, GpuError};
use bytemuck::{Pod, Zeroable};
use pt_core::denoise::DenoiseParams;
use wgpu::util::DeviceExt;

#[repr(C)]
#[derive(Clone, Copy, Pod, Zeroable, Default)]
struct GpuDenoiseParams {
    width: u32,
    height: u32,
    stride: u32,
    sigma_normal: f32,
    sigma_depth: f32,
    sigma_luminance: f32,
    sigma_albedo: f32,
    _pad0: u32,
}

/// Compiled denoise kernels, reusable across frames.
pub struct Denoiser {
    layout: wgpu::BindGroupLayout,
    demodulate: wgpu::ComputePipeline,
    measure: wgpu::ComputePipeline,
    filter: wgpu::ComputePipeline,
    modulate: wgpu::ComputePipeline,
}

impl Denoiser {
    pub fn new(gpu: &Gpu) -> Result<Self, GpuError> {
        let device = &gpu.device;
        let module = gpu.create_shader("denoise/atrous.wgsl")?;
        let sto = |b: u32, read_only: bool| wgpu::BindGroupLayoutEntry {
            binding: b,
            visibility: wgpu::ShaderStages::COMPUTE,
            ty: wgpu::BindingType::Buffer {
                ty: wgpu::BufferBindingType::Storage { read_only },
                has_dynamic_offset: false,
                min_binding_size: None,
            },
            count: None,
        };
        let layout = device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
            label: Some("denoise"),
            entries: &[
                wgpu::BindGroupLayoutEntry {
                    binding: 0,
                    visibility: wgpu::ShaderStages::COMPUTE,
                    ty: wgpu::BindingType::Buffer {
                        ty: wgpu::BufferBindingType::Uniform,
                        has_dynamic_offset: false,
                        min_binding_size: None,
                    },
                    count: None,
                },
                sto(1, true),
                sto(2, false),
                sto(3, false),
                sto(4, false),
                sto(5, false),
            ],
        });
        let pl = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
            label: Some("denoise"),
            bind_group_layouts: &[Some(&layout)],
            ..Default::default()
        });
        let stage = |entry_point: &str| {
            device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
                label: Some(entry_point),
                layout: Some(&pl),
                module: &module,
                entry_point: Some(entry_point),
                compilation_options: Default::default(),
                cache: None,
            })
        };
        Ok(Self {
            layout,
            demodulate: stage("demodulate"),
            measure: stage("measure"),
            filter: stage("atrous_pass"),
            modulate: stage("modulate"),
        })
    }

    /// Filter `accum` into a new buffer of RGBA pixels.
    ///
    /// Ping-pongs `a` and `b` across the à-trous passes by swapping which is
    /// bound where, so the shader always reads `buf_a` and writes `buf_b` and
    /// needs no notion of which pass it is on.
    #[allow(clippy::too_many_arguments)]
    pub fn run(
        &self,
        gpu: &Gpu,
        encoder: &mut wgpu::CommandEncoder,
        accum: &wgpu::Buffer,
        width: u32,
        height: u32,
        params: &DenoiseParams,
        scratch: &DenoiseScratch,
    ) {
        let device = &gpu.device;
        let groups_x = width.div_ceil(8);
        let groups_y = height.div_ceil(8);

        let make_uniform = |stride: u32| {
            device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some("denoise params"),
                contents: bytemuck::bytes_of(&GpuDenoiseParams {
                    width,
                    height,
                    stride,
                    sigma_normal: params.sigma_normal,
                    sigma_depth: params.sigma_depth,
                    sigma_luminance: params.sigma_luminance,
                    sigma_albedo: params.sigma_albedo,
                    _pad0: 0,
                }),
                usage: wgpu::BufferUsages::UNIFORM,
            })
        };
        let group = |u: &wgpu::Buffer, a: &wgpu::Buffer, b: &wgpu::Buffer| {
            device.create_bind_group(&wgpu::BindGroupDescriptor {
                label: Some("denoise"),
                layout: &self.layout,
                entries: &[
                    wgpu::BindGroupEntry {
                        binding: 0,
                        resource: u.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 1,
                        resource: accum.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 2,
                        resource: a.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 3,
                        resource: b.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 4,
                        resource: scratch.deviation.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 5,
                        resource: scratch.output.as_entire_binding(),
                    },
                ],
            })
        };
        let mut dispatch = |pipeline: &wgpu::ComputePipeline, g: &wgpu::BindGroup| {
            let mut p = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
                label: Some("denoise"),
                timestamp_writes: None,
            });
            p.set_pipeline(pipeline);
            p.set_bind_group(0, g, &[]);
            p.dispatch_workgroups(groups_x, groups_y, 1);
        };

        let u0 = make_uniform(1);
        let g0 = group(&u0, &scratch.a, &scratch.b);
        dispatch(&self.demodulate, &g0);
        dispatch(&self.measure, &g0);

        // Track which buffer currently holds the signal.
        let mut in_a = true;
        let mut uniforms = Vec::new();
        let mut groups = Vec::new();
        for level in 0..params.iterations {
            uniforms.push(make_uniform(1 << level));
        }
        for (level, u) in uniforms.iter().enumerate() {
            let _ = level;
            let (src, dst) = if in_a {
                (&scratch.a, &scratch.b)
            } else {
                (&scratch.b, &scratch.a)
            };
            groups.push(group(u, src, dst));
            in_a = !in_a;
        }
        for g in &groups {
            dispatch(&self.filter, g);
        }

        // `modulate` reads `buf_a`, so bind whichever buffer the last pass wrote
        // into that slot.
        let u_final = make_uniform(1);
        let g_final = if in_a {
            group(&u_final, &scratch.a, &scratch.b)
        } else {
            group(&u_final, &scratch.b, &scratch.a)
        };
        dispatch(&self.modulate, &g_final);
    }
}

/// Working buffers, sized to the render target.
pub struct DenoiseScratch {
    a: wgpu::Buffer,
    b: wgpu::Buffer,
    deviation: wgpu::Buffer,
    pub output: wgpu::Buffer,
}

impl DenoiseScratch {
    pub fn new(gpu: &Gpu, pixels: usize) -> Self {
        let zeros4 = vec![0.0f32; pixels.max(1) * 4];
        let zeros1 = vec![0.0f32; pixels.max(1)];
        let device = &gpu.device;
        let rw = |label: &str, bytes: &[u8]| {
            device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some(label),
                contents: bytes,
                // COPY_SRC so the result can be read back — the native harness
                // compares it against the CPU reference.
                usage: wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_SRC,
            })
        };
        Self {
            a: rw("denoise a", bytemuck::cast_slice(&zeros4)),
            b: rw("denoise b", bytemuck::cast_slice(&zeros4)),
            deviation: rw("denoise deviation", bytemuck::cast_slice(&zeros1)),
            // Accum-shaped, so the display pass needs no change. The stride comes
            // from the struct rather than from a literal: it has been 16, 48 and
            // 64 bytes as the guide channels and then the sum of squares landed,
            // and a literal here fails as an out-of-bounds copy in whichever
            // unrelated test reads this buffer next.
            output: rw(
                "denoise output",
                &vec![0u8; pixels.max(1) * std::mem::size_of::<pt_core::gpu_layout::GpuAccum>()],
            ),
        }
    }
}
